### PR TITLE
Allow Porter Operational timeout parameters for `/decrypt`

### LINKED ISSUE #49: Consider increasing learning timeout
Occasionally, I'm getting timeouts:

```
david@Davids-MacBook-Pro-2 nucypher-contracts % curl -X GET <redacted>/get_ursulas -H "Content-Type: application/json" -d '{"quantity": 3}'
{"result": {"failure_message": "Execution stopped before completion - not enough available values (1 failures recorded)", "failures": [{"value": "0xb15d5A4e2be34f4bE154A1b08a94Ab920FfD8A41", "error": "HTTPSConnectionPool(host='<redacted>', port=9151): Read timed out. (read timeout=2)"}]}, "version": "3.1.1"}%  
```

It's possible that current default learning delay of 2 seconds is too short. See:

https://github.com/nucypher/nucypher-porter/blob/799a08ce833bf16dd6c2a867be4e60dba1faf103/porter/main.py#L49

### PR DESCRIPTION
**Pull request by derekpierre**
**Type of PR:**
- [ ] Bugfix
- [X] Feature
- [ ] Documentation
- [ ] Other

**Required reviews:** 
- [ ] 1
- [X] 2
- [ ] 3

**What this does:**
Allow Porter to accept optional timeout values for performing TACo decrypt operation.

While this will be provided, In order for web apps to actually use it, `nucypher-ts`/`taco-api` will need to be updated to allow timeout to optionally specified.

**Issues fixed/closed:**
> - Fixes #...

Fixes #49 .

Depends on `nucypher` PR - https://github.com/nucypher/nucypher/pull/3337

Follow-up for `/reencrypt` filed in #54 .

**Why it's needed:**
> Explain how this PR fits in the greater context of the NuCypher Network.
> E.g., if this PR addresses a `nucypher/productdev` issue, let reviewers know!

**Notes for reviewers:**
> What should reviewers focus on? 
> Is there a particular commit/function/section of your PR that requires more attention from reviewers?

### CODE REVIEW

**Review by piotr-roslaniec — APPROVED** [review]

**Comment by cygnusv** [comment]
Looks good in general, but I'm wondering about potential DoS vectors if we allow timeouts to be defined as input parameters.

**Comment by derekpierre** [comment]
> Looks good in general, but I'm wondering about potential DoS vectors if we allow timeouts to be defined as input parameters.

Great point - we could put a max cap on the timeout? Perhaps that cap can have a default, but be overridden by an env variable? That way the Porter operator can make adjustments to the timeout ... open to other ideas ...

The biggest concern is us simply hardcoding a timeout in the code that is potentially insufficient.

**Comment by cygnusv** [comment]
> > Looks good in general, but I'm wondering about potential DoS vectors if we allow timeouts to be defined as input parameters.
> 
> Great point - we could put a max cap on the timeout? Perhaps that cap can have a default value, but be overridden by an env variable? That way the Porter operator can make adjustments to the timeout ... open to other ideas ...
> 
> Is that something we want to apply to other endpoints eg. `/get_ursulas`?
> 
> The biggest concern is us simply hardcoding a timeout in the code that is potentially insufficient.

Configuration via environment variables or some config file are good. We wouldn't want to make a release just to increment a default timeout.

File: porter/main.py [hunk 1]
```diff
@@ -53,10 +54,14 @@ class Porter(Learner):
     _LONG_LEARNING_DELAY = 30
     _ROUNDS_WITHOUT_NODES_AFTER_WHICH_TO_SLOW_DOWN = 25
 
-    DEFAULT_EXECUTION_TIMEOUT = 15  # 15s
-
     DEFAULT_PORT = 9155
 
+    MAX_GET_URSULAS_TIMEOUT = os.getenv("PORTER_GET_URSULAS_TIMEOUT", default=15)
+    MAX_DECRYPTION_TIMEOUT = os.getenv(
+        "PORTER_MAX_DECRYPTION_TIMEOUT",
```
Comment (cygnusv) [inline porter/main.py:61]: Can envvars be documented somewhere?

File: porter/main.py [hunk 2]
```diff
@@ -171,16 +187,17 @@ def get_ursula_info(ursula_address) -> Porter.UrsulaInfo:
...
-                                 value_factory=value_factory,
-                                 target_successes=quantity,
-                                 timeout=self.execution_timeout,
-                                 stagger_timeout=1)
+        worker_pool = WorkerPool(
+            worker=get_ursula_info,
+            value_factory=value_factory,
+            target_successes=quantity,
+            timeout=timeout,
+            stagger_timeout=1,
```
Comment (cygnusv) [inline porter/main.py:199]: What's this timeout?

**Review by cygnusv — APPROVED** [review]

File: porter/main.py [hunk 3]
```diff
@@ -142,10 +145,23 @@ def _initialize_endpoints(eth_endpoint: str, polygon_endpoint: str):
...
+    ) -> List[UrsulaInfo]:
+        if timeout and timeout > self.MAX_GET_URSULAS_TIMEOUT:
+            self.log.warn(
+                f"Provided sampling timeout ({timeout}s) exceeds "
+                f"maximum ({self.MAX_GET_URSULAS_TIMEOUT}s); "
+                f"using {self.MAX_GET_URSULAS_TIMEOUT}s instead"
+            )
+            timeout = self.MAX_GET_URSULAS_TIMEOUT
+        else:
+            timeout = timeout or self.MAX_GET_URSULAS_TIMEOUT
```
Comment (manumonti) [inline porter/main.py:163]: Not a blocker, but something we can address in the future: if I understand well, we are not checking in the tests that if the provided timeout > MAX_GET_URSULAS_TIMEOUT, the timeout will be MAX_GET_URSULAS_TIMEOUT.

**Review by manumonti — APPROVED** [review]
LGTM! I just left a suggestion

File: porter/main.py (hunk 2 above)
Comment (derekpierre) [reply porter/main.py:199]: This is the timeout between obtaining batches of values for processing - https://github.com/nucypher/nucypher/blob/main/nucypher/utilities/concurrency.py#L322.

In this case batches of staking provider addresses.

File: porter/main.py (hunk 1 above)
Comment (derekpierre) [reply porter/main.py:61]: 👍 I was thinking the threshold docs, and the repos readme.

Those will be done separately.

File: porter/main.py (hunk 3 above)
Comment (derekpierre) [reply porter/main.py:163]: You're right. I'll look into testing that separately. https://github.com/nucypher/nucypher-porter/issues/56

**Pull request merged**